pytest-nunit throwing TypeError: object of type 'ExceptionChainRepr' has no len()

Recently ran into a bug where the tracebacks actually caused n-unit to error when displaying them, so the CI logs were useless until I disabled n-unit output.

Here's the error pytest was throwing after the tests finished running, but no tracebacks or warnings had been displayed:

<details>
<summary> traceback </summary>

```pytb
2021-07-06T06:04:38.6353129Z INTERNALERROR> Traceback (most recent call last):
2021-07-06T06:04:38.6353900Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/main.py", line 269, in wrap_session
2021-07-06T06:04:38.6354460Z INTERNALERROR>     session.exitstatus = doit(config, session) or 0
2021-07-06T06:04:38.6355206Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/main.py", line 323, in _main
2021-07-06T06:04:38.6355702Z INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
2021-07-06T06:04:38.6356458Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
2021-07-06T06:04:38.6356975Z INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
2021-07-06T06:04:38.6357736Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
2021-07-06T06:04:38.6358240Z INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
2021-07-06T06:04:38.6358987Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
2021-07-06T06:04:38.6359788Z INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
2021-07-06T06:04:38.6360624Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
2021-07-06T06:04:38.6361100Z INTERNALERROR>     return outcome.get_result()
2021-07-06T06:04:38.6361812Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
2021-07-06T06:04:38.6362305Z INTERNALERROR>     raise ex[1].with_traceback(ex[2])
2021-07-06T06:04:38.6363020Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
2021-07-06T06:04:38.6363515Z INTERNALERROR>     res = hook_impl.function(*args)
2021-07-06T06:04:38.6364248Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/main.py", line 348, in pytest_runtestloop
2021-07-06T06:04:38.6364817Z INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
2021-07-06T06:04:38.6365579Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
2021-07-06T06:04:38.6366103Z INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
2021-07-06T06:04:38.6366959Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
2021-07-06T06:04:38.6367474Z INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
2021-07-06T06:04:38.6368207Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
2021-07-06T06:04:38.6368784Z INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
2021-07-06T06:04:38.6369588Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
2021-07-06T06:04:38.6370075Z INTERNALERROR>     return outcome.get_result()
2021-07-06T06:04:38.6370845Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
2021-07-06T06:04:38.6371341Z INTERNALERROR>     raise ex[1].with_traceback(ex[2])
2021-07-06T06:04:38.6372066Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
2021-07-06T06:04:38.6372557Z INTERNALERROR>     res = hook_impl.function(*args)
2021-07-06T06:04:38.6373294Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/runner.py", line 109, in pytest_runtest_protocol
2021-07-06T06:04:38.6373820Z INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
2021-07-06T06:04:38.6374553Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/runner.py", line 120, in runtestprotocol
2021-07-06T06:04:38.6375071Z INTERNALERROR>     rep = call_and_report(item, "setup", log)
2021-07-06T06:04:38.6375809Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/runner.py", line 219, in call_and_report
2021-07-06T06:04:38.6376326Z INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
2021-07-06T06:04:38.6377050Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
2021-07-06T06:04:38.6377573Z INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
2021-07-06T06:04:38.6378317Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
2021-07-06T06:04:38.6378832Z INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
2021-07-06T06:04:38.6379564Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
2021-07-06T06:04:38.6380232Z INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
2021-07-06T06:04:38.6381043Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
2021-07-06T06:04:38.6381527Z INTERNALERROR>     return outcome.get_result()
2021-07-06T06:04:38.6382235Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
2021-07-06T06:04:38.6382726Z INTERNALERROR>     raise ex[1].with_traceback(ex[2])
2021-07-06T06:04:38.6383440Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
2021-07-06T06:04:38.6384064Z INTERNALERROR>     res = hook_impl.function(*args)
2021-07-06T06:04:38.6384831Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pytest_nunit/plugin.py", line 293, in pytest_runtest_logreport
2021-07-06T06:04:38.6385364Z INTERNALERROR>     reporter.record_testreport(report)
2021-07-06T06:04:38.6386109Z INTERNALERROR>   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pytest_nunit/plugin.py", line 146, in record_testreport
2021-07-06T06:04:38.6386644Z INTERNALERROR>     if len(testreport.longrepr) > 2:
2021-07-06T06:04:38.6387349Z INTERNALERROR> TypeError: object of type 'ExceptionChainRepr' has no len()
2021-07-06T06:04:38.6387758Z Traceback (most recent call last):
2021-07-06T06:04:38.6388141Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/bin/pytest", line 8, in <module>
2021-07-06T06:04:38.6388513Z     sys.exit(console_main())
2021-07-06T06:04:38.6389184Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185, in console_main
2021-07-06T06:04:38.6389619Z     code = main()
2021-07-06T06:04:38.6390248Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/config/__init__.py", line 162, in main
2021-07-06T06:04:38.6390753Z     ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
2021-07-06T06:04:38.6391441Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
2021-07-06T06:04:38.6391921Z     return self._hookexec(self, self.get_hookimpls(), kwargs)
2021-07-06T06:04:38.6392610Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
2021-07-06T06:04:38.6393081Z     return self._inner_hookexec(hook, methods, kwargs)
2021-07-06T06:04:38.6393756Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
2021-07-06T06:04:38.6394265Z     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
2021-07-06T06:04:38.6395017Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
2021-07-06T06:04:38.6395448Z     return outcome.get_result()
2021-07-06T06:04:38.6396102Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
2021-07-06T06:04:38.6396532Z     raise ex[1].with_traceback(ex[2])
2021-07-06T06:04:38.6397197Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
2021-07-06T06:04:38.6397635Z     res = hook_impl.function(*args)
2021-07-06T06:04:38.6398315Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/main.py", line 316, in pytest_cmdline_main
2021-07-06T06:04:38.6398761Z     return wrap_session(config, _main)
2021-07-06T06:04:38.6399426Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/main.py", line 304, in wrap_session
2021-07-06T06:04:38.6399859Z     config.hook.pytest_sessionfinish(
2021-07-06T06:04:38.6400518Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
2021-07-06T06:04:38.6401074Z     return self._hookexec(self, self.get_hookimpls(), kwargs)
2021-07-06T06:04:38.6401779Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
2021-07-06T06:04:38.6402234Z     return self._inner_hookexec(hook, methods, kwargs)
2021-07-06T06:04:38.6402917Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
2021-07-06T06:04:38.6403431Z     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
2021-07-06T06:04:38.6404180Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 203, in _multicall
2021-07-06T06:04:38.6404591Z     gen.send(outcome)
2021-07-06T06:04:38.6405261Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/_pytest/terminal.py", line 803, in pytest_sessionfinish
2021-07-06T06:04:38.6405692Z     outcome.get_result()
2021-07-06T06:04:38.6406340Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
2021-07-06T06:04:38.6406773Z     raise ex[1].with_traceback(ex[2])
2021-07-06T06:04:38.6407437Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
2021-07-06T06:04:38.6407863Z     res = hook_impl.function(*args)
2021-07-06T06:04:38.6408631Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pytest_nunit/plugin.py", line 368, in pytest_sessionfinish
2021-07-06T06:04:38.6409129Z     full_report = self._create_module_report(self.cases)
2021-07-06T06:04:38.6409849Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pytest_nunit/plugin.py", line 351, in _create_module_report
2021-07-06T06:04:38.6410364Z     [case["stop"] for case in cases.values()], default=datetime.min
2021-07-06T06:04:38.6411070Z   File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/site-packages/pytest_nunit/plugin.py", line 351, in <listcomp>
2021-07-06T06:04:38.6411574Z     [case["stop"] for case in cases.values()], default=datetime.min
2021-07-06T06:04:38.6412022Z KeyError: 'stop'
2021-07-06T06:04:39.3782996Z ##[error]Bash exited with code '1'.
2021-07-06T06:04:39.3795930Z ##[section]Finishing: PyTest
```

</details>

And here's a [gist with the full output once n-unit](https://gist.github.com/ivirshup/48321b8c2282748f874597f7f649d475) was disabled.

The original error was thrown from `numba`, and is about failure to multithread (I believe execution is single threaded here).

-------------

Issue creating the relevant tracebacks: https://github.com/lmcinnes/pynndescent/issues/129

--- a/pytest_nunit/plugin.py
+++ b/pytest_nunit/plugin.py
@@ -151,11 +151,14 @@ class _NunitNodeReporter:
             r["start"] = datetime.utcnow()  # Will be overridden if called
             if testreport.outcome == "skipped":
                 log.debug("skipping : {0}".format(testreport.longrepr))
-                if len(testreport.longrepr) > 2:
+                if isinstance(testreport.longrepr, tuple) and len(testreport.longrepr) > 2:
                     r["error"] = testreport.longrepr[2]
                     r["stack-trace"] = "{0}::{1}".format(
                         testreport.longrepr[0], testreport.longrepr[1]
                     )
+                elif hasattr(testreport.longrepr, "traceback"): # Catches internal ExceptionInfo type
+                    r["error"] = str(testreport.longrepr)
+                    r["stack-trace"] = str(testreport.longrepr.traceback)
                 else:
                     r["error"] = testreport.longrepr
         elif testreport.when == "call":

--- a/pytest_nunit/plugin.py
+++ b/pytest_nunit/plugin.py
@@ -368,10 +368,10 @@ class NunitXML:
         stats["failure"] = outcomes.get("failed", 0)
         stats["skipped"] = outcomes.get("skipped", 0)
         start = min_with_default(
-            [case["start"] for case in cases.values()], default=datetime.min
+            [case["start"] for case in cases.values() if "start" in case], default=datetime.min
         )
         stop = max_with_default(
-            [case["stop"] for case in cases.values()], default=datetime.min
+            [case["stop"] for case in cases.values() if "stop" in case], default=datetime.min
         )
         duration = (stop - start).total_seconds()
         return ModuleReport(

